src/domain/dataverse/aggregate.ts

```diff
@@ -0,0 +1,176 @@
+/* eslint-disable max-lines-per-function */
+import { pipe } from 'fp-ts/function'
+import * as E from 'fp-ts/Either'
+import type * as T from 'fp-ts/Task'
+import * as TE from 'fp-ts/TaskEither'
+import { immer } from 'zustand/middleware/immer'
+import { createStore } from 'zustand/vanilla'
+import type { DataverseDTO } from './dto'
+import type { DataverseEntity } from './entity'
+import type { DataversePort } from './port'
+
+type ByTypeQueryFilter = 'all' | 'DataSpace' | 'Dataset' | 'Service'
+
+type DataverseQueryFilters = {
+  byType: ByTypeQueryFilter
+}
+
+type DataverseQueryError = Error | null
+
+type DataverseQuery = {
+  limit: number
+  hasNext: boolean
+  isLoading: boolean
+  error: DataverseQueryError
+  filters: DataverseQueryFilters
+  language: string
+  setIsLoading: (isLoading: boolean) => void
+  setHasNext: (hasNext: boolean) => void
+  setError: (error: DataverseQueryError) => void
+  setFilters: (newFilter: DataverseQueryFilters) => void
+  setLanguage: (newLng: string) => void
+}
+
+export type DataverseAggregate = {
+  dataverse: DataverseEntity
+  query: DataverseQuery
+  addToDataverse: (newDataverse: DataverseEntity) => void
+  clearDataverse: () => void
+  loadDataverse: () => T.Task<void> | void
```

Comment:
@ccamel 
I can't return only a `T.Task<void>` because the language invariant method `checkLanguageInvariant` returns an `Option<Error>` that is consumed thanks to pattern matching like that: 

```ts
      loadDataverse: (): T.Task<void> | void => {
        const { addToDataverse, query, dataverse } = get()
        const { setIsLoading, setHasNext, setError, limit, language, filters } = query
        const offset = dataverse.length

        setIsLoading(true)
        setError(O.none)

        const checkLanguageInvariant = (): O.Option<Error> =>
          !language.length
            ? O.some(
                new Error(
                  'Oops.. An error occurred when trying to load Dataverse.. Language cannot be empty..'
                )
              )
            : O.none

        const fetchDataverse = (): T.Task<void> =>
          pipe(
            gateway.retrieveDataverse(language, limit, offset, filters),
            TE.match(
              e => {
                setIsLoading(false)
                setError(O.some(e))
              },
              r => {
                setIsLoading(false)
                setHasNext(r.query.hasNext)
                addToDataverse(r.data)
              }
            )
          )

        return pipe(
          checkLanguageInvariant(),
          O.matchW(fetchDataverse, e => setError(O.some(e)))
        )
      }
```

How can we chain the `checkLanguageInvariant` and the `fetchDataverse` to return a `Task` in both cases?

Reply:
your `matchW` is not correct. Use a plain `match` instead.

Something like:

```ts
        return pipe(
          checkLanguageInvariant(),
          O.match(fetchDataverse, e => { setError(O.some(e)); return T.of(undefined) })
        )
```

or

```ts
        return pipe(
          checkLanguageInvariant(), 
          O.match(fetchDataverse, flow(O.some, setError, T.of)))
```

or 

...